File: src/bmsx/emulator/runtime_timing.ts
import { $ } from '../core/engine_core';
import { HZ_SCALE } from '../platform/platform';

export function calcCyclesPerFrameScaled(cpuHz: number, refreshHzScaled: number): number {
	if (!Number.isSafeInteger(cpuHz) || cpuHz <= 0) {
		throw new Error('[RuntimeTiming] cpuHz must be a positive safe integer.');
	}
	if (!Number.isSafeInteger(refreshHzScaled) || refreshHzScaled <= 0) {
		throw new Error('[RuntimeTiming] refreshHzScaled must be a positive safe integer.');
	}
	const wholeCycles = Math.floor(cpuHz / refreshHzScaled) * HZ_SCALE;
	const remainderCycles = Math.floor((cpuHz % refreshHzScaled) * HZ_SCALE / refreshHzScaled);
	const cyclesPerFrame = wholeCycles + remainderCycles;
	if (!Number.isSafeInteger(cyclesPerFrame) || cyclesPerFrame <= 0) {
		throw new Error('[RuntimeTiming] cycles per frame must be a positive safe integer.');
	}
	return cyclesPerFrame;
}

export function resolveVblankCycles(cpuFreqHz: number, ufpsScaled: number, renderHeight: number): number {
	if (!Number.isSafeInteger(cpuFreqHz) || cpuFreqHz <= 0) {
		throw new Error('[RuntimeTiming] cpuFreqHz must be a positive safe integer.');
	}
	if (!Number.isSafeInteger(ufpsScaled) || ufpsScaled <= 0) {
		throw new Error('[RuntimeTiming] ufpsScaled must be a positive safe integer.');
	}
	if (!Number.isSafeInteger(renderHeight) || renderHeight <= 0) {
		throw new Error('[RuntimeTiming] renderHeight must be a positive safe integer.');
	}
	const cycleBudgetPerFrame = calcCyclesPerFrameScaled(cpuFreqHz, ufpsScaled);
	const activeScanlines = Math.floor(cycleBudgetPerFrame / (renderHeight + 1));
	const activeDisplayCycles = activeScanlines * renderHeight;
	const vblankCycles = cycleBudgetPerFrame - activeDisplayCycles;
	if (!Number.isSafeInteger(vblankCycles) || vblankCycles < 0 || vblankCycles > cycleBudgetPerFrame) {
		throw new Error('[RuntimeTiming] invalid vblank cycle configuration.');
	}
	return vblankCycles;
}

export function resolveUfpsScaled(value: number | undefined): number {
	if (value === undefined) {
		throw new Error('[RuntimeTiming] machine.ufps is required.');
	}
	if (!Number.isSafeInteger(value) || value <= HZ_SCALE) {
		throw new Error('[RuntimeTiming] machine.ufps must be a safe integer greater than 1 Hz.');
	}
	return value;
}

export class RuntimeTimingState {
	public ufpsScaled = 0;
	public targetFps = 0;
	public frameDurationMs = 0;

	constructor(ufpsScaled: number) {
		this.applyUfpsScaled(ufpsScaled);
	}

	public get ufps(): number {
		return this.targetFps;
	}

	public applyUfpsScaled(ufpsScaled: number): void {
		this.ufpsScaled = resolveUfpsScaled(ufpsScaled);
		this.targetFps = this.ufpsScaled / HZ_SCALE;
		this.frameDurationMs = 1000 / this.targetFps;
		$.platform.audio.setFrameTimeSec(HZ_SCALE / this.ufpsScaled);
		$.sndmaster.setMixerFps(this.targetFps);
	}

	public resolveCycleBudget(cpuHz: number): number {
		return calcCyclesPerFrameScaled(cpuHz, this.ufpsScaled);
	}

	public resolveVblankCycles(cpuHz: number, renderHeight: number): number {
		return resolveVblankCycles(cpuHz, this.ufpsScaled, renderHeight);
	}
}

File: src/bmsx_cpp/emulator/runtime_timing.cpp
#include "runtime_timing.h"
#include <stdexcept>

namespace bmsx {

int calcCyclesPerFrame(i64 cpuHz, i64 refreshHzScaled) {
	if (cpuHz <= 0) {
		throw std::runtime_error("[RuntimeTiming] cpuHz must be a positive integer.");
	}
	if (refreshHzScaled <= HZ_SCALE) {
		throw std::runtime_error("[RuntimeTiming] refreshHzScaled must be greater than 1 Hz.");
	}
	const i64 wholeCycles = (cpuHz / refreshHzScaled) * HZ_SCALE;
	const i64 remainderCycles = ((cpuHz % refreshHzScaled) * HZ_SCALE) / refreshHzScaled;
	const i64 cyclesPerFrame = wholeCycles + remainderCycles;
	if (cyclesPerFrame <= 0) {
		throw std::runtime_error("[RuntimeTiming] cycles per frame must be a positive integer.");
	}
	return static_cast<int>(cyclesPerFrame);
}

i64 resolveVblankCycles(i64 cpuHz, i64 refreshHzScaled, i32 renderHeight) {
	if (renderHeight <= 0) {
		throw std::runtime_error("[RuntimeTiming] renderHeight must be a positive integer.");
	}
	const i64 cycleBudgetPerFrame = calcCyclesPerFrame(cpuHz, refreshHzScaled);
	const i64 activeScanlines = cycleBudgetPerFrame / static_cast<i64>(renderHeight + 1);
	const i64 activeDisplayCycles = activeScanlines * static_cast<i64>(renderHeight);
	const i64 vblankCycles = cycleBudgetPerFrame - activeDisplayCycles;
	if (vblankCycles <= 0) {
		throw std::runtime_error("[RuntimeTiming] vblank_cycles must be greater than 0.");
	}
	if (vblankCycles > cycleBudgetPerFrame) {
		throw std::runtime_error("[RuntimeTiming] vblank_cycles must be less than or equal to cycles_per_frame.");
	}
	return vblankCycles;
}

RuntimeTimingState::RuntimeTimingState(i64 ufpsScaledValue) {
	applyUfpsScaled(ufpsScaledValue);
}

void RuntimeTimingState::applyUfpsScaled(i64 value) {
	if (value <= HZ_SCALE) {
		throw std::runtime_error("[RuntimeTiming] machine.ufps must be greater than 1 Hz.");
	}
	ufpsScaled = value;
	ufps = static_cast<f64>(ufpsScaled) / static_cast<f64>(HZ_SCALE);
	frameDurationMs = 1000.0 / ufps;
}

} // namespace bmsx
